frame/state-trie-migration/src/lib.rs

```diff
@@ -0,0 +1,1514 @@
+			ensure!(T::Currency::can_slash(&who, deposit), "not enough funds");
+
+			let mut task = Self::migration_process();
+			task.migrate_until_exhaustion(limits);
+
+			// ensure that the migration witness data was correct.
+			if real_size != task.dyn_size {
+				// let the imbalance burn.
+				let (_imbalance, _remainder) = T::Currency::slash(&who, deposit);
+				debug_assert!(_remainder.is_zero());
+				return Err("Wrong witness data".into())
+			}
+
+			Self::deposit_event(Event::<T>::Migrated(
+				task.dyn_top_items,
+				task.dyn_child_items,
+				MigrationCompute::Signed,
+			));
+
+			MigrationProcess::<T>::put(task);
+			Ok(Pays::No.into())
+		}
+
+		/// Migrate the list of top keys by iterating each of them one by one.
+		///
+		/// This does not affect the global migration process tracker ([`MigrationProcess`]), and
+		/// should only be used in case any keys are leftover due to a bug.
+		#[pallet::weight(Pallet::<T>::dynamic_weight(keys.len() as u32, *total_size))]
+		pub fn migrate_custom_top(
+			origin: OriginFor<T>,
+			keys: Vec<Vec<u8>>,
+			total_size: u32,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			// ensure they can pay more than the fee.
+			let deposit = T::SignedDepositPerItem::get().saturating_mul((keys.len() as u32).into());
+			ensure!(T::Currency::can_slash(&who, deposit), "not enough funds");
+
+			let mut dyn_size = 0u32;
+			for key in &keys {
+				if let Some(data) = sp_io::storage::get(&key) {
+					dyn_size = dyn_size.saturating_add(data.len() as u32);
+					sp_io::storage::set(key, &data);
+				}
+			}
+
+			if dyn_size != total_size {
+				let (_imbalance, _remainder) = T::Currency::slash(&who, deposit);
+				debug_assert!(_remainder.is_zero());
+				return Err("Wrong witness data".into())
+			}
+
+			Self::deposit_event(Event::<T>::Migrated(
+				keys.len() as u32,
+				0,
+				MigrationCompute::Signed,
+			));
+			Ok(().into())
+		}
+
+		/// Migrate the list of child keys by iterating each of them one by one.
+		///
+		/// All of the given child keys must be present under one `top_key`.
+		///
+		/// This does not affect the global migration process tracker ([`MigrationProcess`]), and
+		/// should only be used in case any keys are leftover due to a bug.
+		#[pallet::weight(Pallet::<T>::dynamic_weight(child_keys.len() as u32, *total_size))]
+		pub fn migrate_custom_child(
+			origin: OriginFor<T>,
+			top_key: Vec<u8>,
+			child_keys: Vec<Vec<u8>>,
+			total_size: u32,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			// ensure they can pay more than the fee.
+			let deposit =
+				T::SignedDepositPerItem::get().saturating_mul((child_keys.len() as u32).into());
+			ensure!(T::Currency::can_slash(&who, deposit), "not enough funds");
+
+			let mut dyn_size = 0u32;
+			for child_key in &child_keys {
+				if let Some(data) =
+					sp_io::default_child_storage::get(Self::child_io_key(&top_key), &child_key)
```

Comment:
`child_io_key` requires trusted data, but here user can call it with a wrong `top_key`
As far as I can see, this allow to the caller to halt the auto/unsigned migration :-/

Can we make `child_io_key` return an result ?
Or at least prepend it with unsafe and add a note before each call.

Reply:
This transaction should not affect the auto/unsigned migration at all. It is merely an entry point to read keys on chain and migrate them. 

The fix is that we should not halt the auto/unsigned migration if this is being called from this transaction, which can be achieved with a wrapper.